core/src/main/java/org/jruby/truffle/runtime/util/ArrayUtils.java

```diff
@@ -17,6 +18,30 @@
 
 public abstract class ArrayUtils {
 
+    /**
+     * Extracts part of an array into a newly allocated Object[] array. Does not perform safety checks on parameters.
+     * @param source the source array whose values should be extracted
+     * @param start the start index, must be >= 0 and < source.length
+     * @param end the end index (exclusive), must be >= 0 and <= source.length and >= start
+     * @return a newly allocated array with the extracted elements and length (end - start)
+     */
+    public static Object[] extractRange(Object[] source, int start, int end) {
+        assert checkExtractRangeArgs(source, start, end);
+        int length = end - start;
+        Object[] result = new Object[length];
+        System.arraycopy(source, start, result, 0, length);
+        return result;
+    }
```

Comment:
The method Arrays.copyOfRange is not suitable for partial evaluation. It contains an exception branch with String concatenation. While we usually expect this branch to be optimized away, we cannot guarantee that HotSpot profiling gives the expected results (because completely different code may execute and then "enable" the path). Additionally, the semantics of extractRange is more complicated than what we usually need - i.e., it will return an array of the same type as the source array, but we only need it to be an Object[] array. Overall, this is again a small change aimed at reducing startup/warmup times - the part about the exception branch can also have peak performance impacts.